tools/pre-commit

```diff
@@ -1 +1,24 @@
-./tools/lint-all $(git diff --cached --name-only --diff-filter=ACM) || true
+#!/bin/sh
+
+# This hook runs the Zulip code linter ./tools/lint-all and returns true
+# regardless of linter results so that your commit may continue.
+
+# Messages from the linter will output to screen.
+#
+# If you are using the vagrant method for your Zulip dev setup, the linter will
+# be run through vagrant ssh.
+
+if `which vagrant > /dev/null` && vstatus="$(vagrant status)" && ! [[ $vstatus =~ 'not created' ]]
+then
+  vcmd='/srv/zulip/tools/lint-all $(cd /srv/zulip && git diff --cached --name-only --diff-filter=ACM) || true'
+  if ! [[ $vstatus =~ 'running' ]]
+  then
+    echo "Vagrant machine not running; lint-all skipped."
+  else
+    echo "Running lint-all using vagrant..."
+    vagrant ssh -c "$vcmd" 2> /dev/null
+  fi
+else
+  echo "Running lint-all..."
+  ./tools/lint-all $(git diff --cached --name-only --diff-filter=ACM) || true
+fi
```

Comment:
This implementation is going to substantially reduce `lint-all` performance in contexts where Vagrant is installed but `lint-all` would run locally without it.  I think we want the ordering of the logic to be this:

(1) Check `[ -e ../.vagrant ]`, `which vagrant`, whether the username is Vagrant, and whether the user is not in a virtualenv.  All of these tests take at most a few milliseconds, and if they fail, we should try running the local `lint-all`, since either Vagrant definitely won't work (as with the first first few tests), or `lint-all` will run much faster not using Vagrant.  

(2) Only if all of those suggest Vagrant is installed do we try running a Vagrant command.  They are all very slow, e.g.:
```
$ time vagrant status
A Vagrant environment or target machine is required to run this
command. Run `vagrant init` to create a new Vagrant environment. Or,
get an ID of a target machine from `vagrant global-status` to run
this command on. A final option is to change to a directory with a
Vagrantfile and to try again.

real	0m0.809s
user	0m0.756s
sys	0m0.044s
```

So we might want to think about whether we want to do that check at all or just try doing `vagrant ssh` once we've confirmed .